File: SDS/graph2/graph2_A_11266.cpp
#include<iostream>
#include<vector>

#define MAX_V 10001

using namespace std;

bool iscut[MAX_V];
int order[MAX_V];
vector<int> graph[MAX_V];

int count = 0;

int find(int here, bool root)
{
    ++count;
    order[here] = count;
    
    int minOrder = order[here];
    int childCount = 0;
    
    for(int i=0; i < graph[here].size(); i++)
    {
        int there = graph[here][i];
        
        if(order[there])
        {
            minOrder = min(minOrder, order[there]);
            continue;
        }
        
        childCount += 1;
        
        int prev = find(there, false);
        
        if(!root && prev >= order[here])
            iscut[here] = true;
            
        minOrder = min(minOrder, prev);
    }
    
    if(root && childCount >= 2)
        iscut[here] = true;
        
    return minOrder;
            
}

int main()
{
    int V, E;
    
    scanf("%d %d", &V, &E);
    
    for(int e=0; e < E; e++)
    {
        int f, t;
        scanf("%d %d", &f, &t);
        
        graph[f].push_back(t);
        graph[t].push_back(f);
    }
    
    for(int i=1; i <= V; i++)
        if(order[i] == 0)
            find(i, true);
            
    int c = 0;
    for(int i=1; i <= V; i++)
        if(iscut[i])
            c += 1;
            
    printf("%d\n", c);
    
    for(int i=1; i <= V; i++)
        if(iscut[i])
            printf("%d ", i);
            
    printf("\n");
}

File: SDS/graph2/graph2_A_11266.py
from sys import stdin
V, E = map(int, stdin.readline().split())

graph = [ [] for _ in range(V+1)]
order = [ 0 for _ in range(V+1) ]
iscut = [ False for _ in range(V+1)]

count = 0

def find(here, is_root):
    global count 
    
    count += 1
    order[here] = count
    
    
    child_count = 0
    
    min_order = order[here]
    
    for child in graph[here]:
        if order[child] != 0:
            min_order = min(min_order, order[child])
            continue
        
        child_count += 1
        
        prev = find(child, False)
            
        if not is_root and prev >= order[here]:
            iscut[here] = True
            
        min_order = min(prev, min_order)
    
    if is_root:
        if child_count >= 2:
            iscut[here] = True
    
    return min_order

for e in range(E):
    f, t = map(int, stdin.readline().split())
    graph[f].append(t)
    graph[t].append(f)
    
for i in range(1, V+1):
    if order[i] == 0:
        find(i, True)

print(iscut.count(True))

result = []
for i in range(1, V+1):
    if iscut[i]:
        result.append(str(i))
print(' '.join(result))

